peer.go

```diff
@@ -425,6 +425,29 @@ func (p *peer) Start() error {
 		}
 	}
 
+	// Node announcements don't propagate very well throughout the network
+	// as there isn't a way to efficiently query for them through their
+	// timestamp, mostly affecting nodes that were offline during the time
+	// of broadcast. We'll resend our node announcement to them as a
+	// best-effort delivery such that it can also propagate to their peers.
+	//
+	// TODO(wilmer): Remove this once we're able to query for node
+	// announcements through their timestamps.
+	if len(activeChans) > 0 {
```

Comment:
i think there are cases where this would violate the spec, bc we might send it too early.

we want to send this only in the case where:
 - we have 1+ non-pending, public channels w/ this peer, and
 - we have already exchanged channel announcements,
 - and we've already exchanged channel updates

is it possible to have the node announcement sent after anything we already have queued in the reliable sender? i think that would check the boxes. even tho the reliable sender already violates portions of the spec, i think it would be more in-line at least wrt to ordering of gossip messages.

Reply:
IIRC node announcements are only dependent on a channel announcement, the channel update is independent, though they are ordered in the fashion you described within gossip queries.

The rationale was that peers would just ignore the node announcement, but it would have some negative implications if some kind of banning heuristic was in place. With the current diff in place, a violation would occur when we've only ever had pending channels with the connected peer and no confirmed channels with any others, which we could somewhat cover by extending the if condition.

> even tho the reliable sender already violates portions of the spec

In what ways?

Reply:
> IIRC node announcements are only dependent on a channel announcement, the channel update is independent, though they are ordered in the fashion you described within gossip queries.

Double checked with the spec and you're correct, if the receiver doesn't know of the channel_announcement it SHOULD just ignore the node_announcement. Great, that's better than what I had recalled

> With the current diff in place, a violation would occur when we've only ever had pending channels with the connected peer and no confirmed channels with any others, which we could somewhat cover by extending the if condition.

So we are guaranteed to violate on the first channel we create, and with each new channel until a one confirms? That sounds tolerable on our end, I wonder how other implementations would handle this.

Also this diff looks like it will try to send a node announcement even if all of the sender's channels are private, is there something else preventing that?

> In what ways?

The spec has changed a bit since this was implemented, but currently it requires the following ordering:
 - `ChannelReestablish` [-> `FundingLocked`] [-> `ChannelAnnouncement`] [-> `ChannelUpdate`]

Because the reliable sender isn't integrated with the link, we have two parallel, unsynchronized processes:
 - `ChannelReestablish` [-> `FundingLocked`]
 - [`ChannelAnnouncement`] [-> `ChannelUpdate`]

which can result in various interleavings that are invalid according to the spec. Note it was invalid before the spec changed, but the recent modifications have made it more invalid.

Relevant modifications from the spec:

> - MUST NOT send announcement_signatures messages until funding_locked has been sent and received AND the funding transaction has at least six confirmations
> - MUST NOT send a created channel_update before funding_locked has been received.

To make matters worse it's also possible for lnd to send a duplicate `FundingLocked` before `ChannelReestablish`. There is a multi-step plan to resolve all of these issues, of which #4092 was phase one.

Reply:
> So we are guaranteed to violate on the first channel we create, and with each new channel until a one confirms? That sounds tolerable on our end, I wonder how other implementations would handle this.
Also this diff looks like it will try to send a node announcement even if all of the sender's channels are private, is there something else preventing that?

I can push an update to address these if we happen to proceed with this approach. We'd make sure there is at least one confirmed public channel before sending it. The only violation in this case would be if either peer restarts after the channel has confirmed and before `FundingLocked` and `ChannelAnnouncement` have been processed. After they have been processed, there shouldn't be any violations in subsequent restarts.

If we happen to address the issues with the reliable sender, then I would be fine integrating this there, though it would require a bit more work than the current approach, and this is just meant to be temporary to help node announcements propagate better until we're able to query them through their timestamps.

Reply:
> I can push an update to address these if we happen to proceed with this approach. We'd make sure there is at least one confirmed public channel before sending it.

Sounds good, I think that's the right approach.

> If we happen to address the issues with the reliable sender, then I would be fine integrating this there, though it would require a bit more work than the current approach, and this is just meant to be temporary to help node announcements propagate better until we're able to query them through their timestamps.

Wouldn't worry about adding more things to the reliable sender, that logic will end up being a part of the links when all is said and done.

Reply:
> Sounds good, I think that's the right approach.

Done!